sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -162,9 +475,127 @@ public void output(OutputT output, Instant timestamp, BoundedWindow window) {
             outputTo(consumers, WindowedValue.of(output, timestamp, window, PaneInfo.NO_FIRING));
           }
         };
+    switch (context.pTransform.getSpec().getUrn()) {
+      case PTransformTranslation.SPLITTABLE_SPLIT_RESTRICTION_URN:
+        this.outputSplitRestrictionReceiver =
+            new OutputReceiver<RestrictionT>() {
+
+              @Override
+              public void output(RestrictionT output) {
+                outputTo(
+                    mainOutputConsumers,
+                    (WindowedValue<OutputT>)
+                        currentElement.withValue(KV.of(currentElement.getValue(), output)));
+              }
+
+              @Override
+              public void outputWithTimestamp(RestrictionT output, Instant timestamp) {
+                outputTo(
+                    mainOutputConsumers,
+                    (WindowedValue<OutputT>)
+                        WindowedValue.of(
+                            KV.of(currentElement.getValue(), output),
+                            timestamp,
+                            currentWindow,
+                            currentElement.getPane()));
+              }
+            };
+        break;
+      case PTransformTranslation.SPLITTABLE_SPLIT_AND_SIZE_RESTRICTIONS_URN:
+        this.outputSplitRestrictionReceiver =
+            new OutputReceiver<RestrictionT>() {
+
+              @Override
+              public void output(RestrictionT output) {
+                RestrictionTracker<RestrictionT, PositionT> outputTracker =
+                    doFnInvoker.invokeNewTracker(output);
+                outputTo(
+                    mainOutputConsumers,
+                    (WindowedValue<OutputT>)
+                        currentElement.withValue(
+                            KV.of(
+                                KV.of(currentElement.getValue(), output),
+                                outputTracker instanceof HasSize
+                                    ? ((HasSize) outputTracker).getSize()
+                                    : 1.0)));
+              }
+
+              @Override
+              public void outputWithTimestamp(RestrictionT output, Instant timestamp) {
+                outputTo(
+                    mainOutputConsumers,
+                    (WindowedValue<OutputT>)
+                        WindowedValue.of(
+                            KV.of(currentElement.getValue(), output),
```

Comment:
Yes, fixed in the version below.

I'm relying on the migration to using SDF everywhere will catch the edge cases via the validates runner tests since unit testing the execution of a single instance is quite verbose.